### PR TITLE
Update for client IP settings

### PR DESCRIPTION
**Pull request by nikolaystoilov-okta**
- Added sample of vendor specific dictionary and custom attributes (PaloAlto).
- Added additional parameters for vendor specific attributes - vendor ID and vendor type.
- Added parameter for question as second factor authentication.
- Added processing of mfa selection.

### CODE REVIEW

File: src/com/okta/SmsAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+public class SmsAuthenticator extends PAPAuthenticator {
+    private RadiusClient radiusClient;
+
+    public SmsAuthenticator(RadiusClient rc) {
+        radiusClient = rc;
+    }
+
+    @Override
+    public void processChallenge(RadiusPacket request, RadiusPacket challenge) throws RadiusException {
```
Comment (nandagopalseshagiri-okta) [inline src/com/okta/SmsAuthenticator.java]: Why is this separation needed for each factor? The code currently supports all 2nd factor flow - there is no need to write extra code for each factor type.

Comment (nikolaystoilov-okta) [reply src/com/okta/SmsAuthenticator.java]: Updated. I wanted to be able to supply the answers to security question from CLI.

File: src/com/okta/ConsoleClient.java [hunk 2]
```diff
@@ -177,17 +218,29 @@ public static void main(String[] args) {
...
-            RadiusClient rc = new RadiusClient(host, argMap.get(Params.secret), hp.v, 1813, 10);
+            RadiusClient rc = new RadiusClient(host, argMap.get(Params.secret), hp.v, radiusAccountServerPort, timeOut);
 
             RadiusRequest request = new AccessRequest(rc, getBaseAttrList(argMap));
             request.addAttribute(new Attr_UserPassword(argMap.get(Params.password)));
 
             System.out.println("Sending:\n" + request.toString());
 
-            RadiusResponse reply = rc.authenticate((AccessRequest) request, new SmsAuthenticator(rc), 5);
+            RadiusAuthenticator authenticator = null;
```
Comment (nandagopalseshagiri-okta) [inline src/com/okta/ConsoleClient.java]: Though SmsAuthenticator was a poor name - it was handling the factor selection and all other factors, not just SMS, correctly. So I am wondering what is the need for DefaultAuthenticator, MfaSelectorAuthenticator and QuestionAuthenticator.

File: src/com/okta/ConsoleClient.java [hunk 3]
```diff
@@ -107,39 +127,72 @@ public static void usage() {
...
+        return attrs;
+    }
 
-                this.username = null;
-                this.password = null;
+    private static void addStandardAttributes(Map<Params, String> argMap, AttributeList attrs, String clientIp) {
+        String standardCode = argMap.get(Params.standardCode);
+        if (standardCode != null) {
+            long code = Integer.parseInt(standardCode);
+            if (code < 1  && code > 64) {
```
Comment (nandagopalseshagiri-okta) [inline src/com/okta/ConsoleClient.java]: you mean || in the place of &&

Comment (nikolaystoilov-okta) [reply src/com/okta/ConsoleClient.java]: updated.

File: src/com/okta/Attr_PaloAltoClientSourceIP.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public static final long TYPE = 26L;
+    public static final long serialVersionUID = 7L;
+
+    public void setup() {
+        this.attributeName = NAME;
+        this.attributeType = TYPE;
+        this.vendorId = VENDOR_ID;
+        this.vsaAttributeType = VSA_TYPE;
+        this.attributeValue = new StringValue();
+        //this.attributeValue = new IPAddrValue();
```
Comment (nandagopalseshagiri-okta) [inline src/com/okta/Attr_PaloAltoClientSourceIP.java]: :nit remove this line

**Comment by nandagopalseshagiri-okta** [comment]
Looks good. Go ahead and merge.

**Pull request merged**